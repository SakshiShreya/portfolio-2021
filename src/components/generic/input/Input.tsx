import { FunctionComponent } from "react";
import styles from "./Input.module.scss";

interface CommonProps {
  label: string;
}

interface InputProps extends CommonProps, React.InputHTMLAttributes<HTMLInputElement> {
  type?: "text" | "email";
  forwardedRef?: React.RefObject<HTMLInputElement>;
}

interface TextAreaProps extends CommonProps, React.TextareaHTMLAttributes<HTMLTextAreaElement> {
  type: "textarea";
  forwardedRef?: React.RefObject<HTMLTextAreaElement>;
}

const Input: FunctionComponent<InputProps | TextAreaProps> = ({ type, label, id, forwardedRef, ...props }) => {
  return (
    <div className={styles.group}>
      {type === "textarea" ? (
        <textarea
          placeholder={label}
          className={[styles.field, styles.textarea].join(" ")}
          id={id}
          ref={forwardedRef as TextAreaProps["forwardedRef"]}
          {...(props as React.TextareaHTMLAttributes<HTMLTextAreaElement>)}
        />
      ) : (
        <input
          type={type}
          placeholder={label}
          className={styles.field}
          id={id}
          ref={forwardedRef as InputProps["forwardedRef"]}
          {...(props as React.InputHTMLAttributes<HTMLInputElement>)}
        />
      )}
      <label htmlFor={id} className={styles.label}>
        {label}
      </label>
    </div>
  );
};

export default Input;
