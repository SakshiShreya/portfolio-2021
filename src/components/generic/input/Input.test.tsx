import { render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Input from "./Input";

// no need to check other functionalities because those are handled by default input element
it("should render Input", () => {
  const elem = renderer.create(<Input label="test" />).toJSON();
  expect(elem).toMatchSnapshot();
});

describe("test input for type", () => {
  it("should render input of type text if text is passed", () => {
    const { container } = render(<Input label="Test" type="text" />);
    expect(container.querySelector("input")).toBeDefined();
    expect(container.querySelector("input")).toHaveAttribute("type", "text");
  });

  it("should render input of type email if email is passed", () => {
    const { container } = render(<Input label="Test" type="email" />);
    expect(container.querySelector("input")).toBeDefined();
    expect(container.querySelector("input")).toHaveAttribute("type", "email");
  });

  it("should render textarea if textarea is passed", () => {
    const { container } = render(<Input label="Test" type="textarea" />);
    expect(container.querySelector("textarea")).toBeDefined();
  });
});

it("should have correct label", () => {
  const { container } = render(<Input label="Test" />);
  expect(container.querySelector("label")).toBeDefined();
  expect(container.querySelector("label")).toHaveTextContent("Test");
});
