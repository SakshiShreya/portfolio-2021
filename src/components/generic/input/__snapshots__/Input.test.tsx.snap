// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Input 1`] = `
<div
  className="group"
>
  <input
    className="field"
    placeholder="test"
  />
  <label
    className="label"
  >
    test
  </label>
</div>
`;
